### PR TITLE
add `SetSubnetOwner` to `Chain` interface

### PR DESCRIPTION
**Pull request by dhrubabasu**

### CODE REVIEW

File: vms/platformvm/state/diff.go [hunk 1]
```diff
@@ -562,5 +566,8 @@ func (d *diff) Apply(baseState State) error {
 			baseState.DeleteUTXO(utxoID)
 		}
 	}
+	for subnetID, owner := range d.subnetOwners {
+		baseState.SetSubnetOwner(subnetID, owner)
+	}
```
Comment (StephenButtolph) [inline vms/platformvm/state/diff.go:571]: This is a really weird asymmetry. If I call `AddSubnet` on a `Diff` that will call `AddSubnet` and `SetSubnetOwner` on the base state.

File: vms/platformvm/state/state.go [hunk 2]
```diff
@@ -816,9 +850,45 @@ func (s *state) AddSubnet(createSubnetTx *txs.Tx) {
...
+		}
+
+		s.SetSubnetOwner(subnetID, owner)
+		s.subnetOwnerCache.Put(subnetID, &wrappedFxOwner{
+			owner: owner,
+			size:  len(ownerBytes),
+		})
+
+		return owner, nil
+	}
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:890]:
```suggestion
	if err == nil {
		var owner fx.Owner
		if _, err := block.GenesisCodec.Unmarshal(ownerBytes, &owner); err != nil {
			return nil, err
		}

		s.SetSubnetOwner(subnetID, owner)
		s.subnetOwnerCache.Put(subnetID, &wrappedFxOwner{
			owner: owner,
			size:  len(ownerBytes),
		})

		return owner, nil
	}
	if err != database.ErrNotFound {
		return nil, err
	}
```

File: vms/platformvm/state/state.go [hunk 3]
```diff
@@ -816,9 +850,45 @@ func (s *state) AddSubnet(createSubnetTx *txs.Tx) {
...
+
 	subnetIntf, _, err := s.GetTx(subnetID)
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:892]: I think we should be caching the miss here right?

File: vms/platformvm/state/state.go [hunk 4]
```diff
@@ -834,9 +904,20 @@ func (s *state) GetSubnetOwner(subnetID ids.ID) (fx.Owner, error) {
...
 }
 
+func (s *state) SetSubnetOwner(subnetID ids.ID, owner fx.Owner) {
+	if s.subnetOwners == nil {
+		s.subnetOwners = map[ids.ID]fx.Owner{
+			subnetID: owner,
+		}
+	} else {
+		s.subnetOwners[subnetID] = owner
+	}
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:918]: We never `nil` out `subnetOwners` - so we should probably just initialize it once and then remove this check

File: vms/platformvm/state/state.go [hunk 5]
```diff
@@ -2272,6 +2354,25 @@ func (s *state) writeSubnets() error {
 	return nil
 }
 
+func (s *state) writeSubnetOwners() error {
+	for subnetID, owner := range s.subnetOwners {
+		subnetID := subnetID
+		owner := owner
+		delete(s.subnetOwners, subnetID)
+		s.subnetOwnerCache.Evict(subnetID) // TODO: Should this be Evict?
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:2362]: I think we can just cache this here rather than evicting

File: vms/platformvm/state/diff.go (hunk 1 above)
Comment (dhrubabasu) [reply vms/platformvm/state/diff.go:567]: Adjusted it to be explicitly set when calling `AddSubnet`

File: vms/platformvm/state/state.go [hunk 6]
```diff
@@ -420,6 +429,18 @@ type txAndStatus struct {
...
+	owner fx.Owner
+	size  int
+}
+
+func wrappedFxOwnerSize(_ ids.ID, w *wrappedFxOwner) int {
+	if w == nil {
+		return ids.IDLen + constants.PointerOverhead
+	}
+	return ids.IDLen + w.size + constants.PointerOverhead
+}
```
Comment (abi87) [inline vms/platformvm/state/state.go:442]: I understand this wrapper is just to be able to use fx.Owner in a sizedCache without recalculating the size every time.
The wrapper feels like a band-aid to me. I'd add a size method to fx.Owner interface to let ever owner be used in a sizedCache. Also fx.Owner implementation may be able to calculate statically their serialized size, without need to actually serialize data

**Review by abi87 — CHANGES_REQUESTED** [review]
I'd handle fx.Owner size calculation differently.
Moreover I believe this change would require a db migration. Will it be part of a different PR?

File: vms/platformvm/state/state.go (hunk 6 above)
Comment (dhrubabasu) [reply vms/platformvm/state/state.go:442]: I didn't think it made sense to have a `Bytes()` like method for `fx.Owner` since the normal usage is passing it into another struct. This is a special case since we're explicitly tracking the owners separately so I'd like to keep it contained here

**Comment by dhrubabasu** [comment]
> Moreover I believe this change would require a db migration. Will it be part of a different PR?

This will not require a db migration, I'm lazily populating the subnetOwner db.

**Review by abi87 — CHANGES_REQUESTED** [review]
I'd introduce `Size()` method for` fx.Owner` so to avoid complicating P-chain state further.

File: vms/platformvm/state/state.go [hunk 7]
```diff
@@ -352,6 +356,11 @@ type state struct {
 	subnetBaseDB  database.Database
 	subnetDB      linkeddb.LinkedDB
 
+	// Subnet ID --> Owner of the subnet
+	subnetOwners     map[ids.ID]fx.Owner
+	subnetOwnerCache cache.Cacher[ids.ID, *wrappedFxOwner] // cache of subnetID -> owner if the entry is nil, it is not in the database
```
Comment (joshua-kim) [inline vms/platformvm/state/state.go:361]: feel free to ignore this comment: It's consistent w/ the rest of the code but I personally don't think we need comments here.

File: vms/platformvm/state/state.go [hunk 8]
```diff
@@ -420,6 +429,18 @@ type txAndStatus struct {
 	status status.Status
 }
 
+type wrappedFxOwner struct {
+	owner fx.Owner
+	size  int
+}
+
+func wrappedFxOwnerSize(_ ids.ID, w *wrappedFxOwner) int {
```
Comment (joshua-kim) [inline vms/platformvm/state/state.go:437]: If we change our cache to be of type `wrappedFxOwner` instead of `*wrappedFxOwner` I think we can simplify the below if to be the following:

```
return ids.IDLen + w.size
```

And make the invariant that if `wrappedFxOwner` is the zero value `wrappedFxOwner{}`, then the key is not in the database.

File: vms/platformvm/state/state.go [hunk 9]
```diff
@@ -420,6 +429,18 @@ type txAndStatus struct {
 	status status.Status
 }
 
+type wrappedFxOwner struct {
```
Comment (joshua-kim) [inline vms/platformvm/state/state.go:432]: I wanna name this something else... just not huge on using `wrapped*` as a name since it's a little hard for me to know what this is without looking at the code closely (I had to read the surrounding cache code to figure out what this was doing). Can we call this something like `subnetOwnerSize` or anything along the lines of `*ownerSize`?

File: vms/platformvm/state/state.go (hunk 6 above)
Comment (joshua-kim) [reply vms/platformvm/state/state.go:442]: I would prefer implementing this wrapper for now and if we find that we keep needing this in other places we can consider implementing it on the `fx` package. I like package being able to define their own abstractions and keeping the interface as minimal as possible

File: vms/platformvm/state/state.go [hunk 10]
```diff
@@ -573,6 +594,16 @@ func newState(
 
 	subnetBaseDB := prefixdb.New(subnetPrefix, baseDB)
 
+	subnetOwnerDB := prefixdb.New(subnetOwnerPrefix, baseDB)
+	subnetOwnerCache, err := metercacher.New[ids.ID, *wrappedFxOwner](
```
Comment (joshua-kim) [inline vms/platformvm/state/state.go:598]: ref earlier comment

File: vms/platformvm/state/state.go [hunk 11]
```diff
@@ -2272,6 +2342,29 @@ func (s *state) writeSubnets() error {
 	return nil
 }
 
+func (s *state) writeSubnetOwners() error {
+	for subnetID, owner := range s.subnetOwners {
+		subnetID := subnetID
+		owner := owner
```
Comment (joshua-kim) [inline vms/platformvm/state/state.go:2348]: I thought loop variables in go were created once and the value was updated with each iteration, so I thought we wouldn't need these copies here. I think it's safe to remove these lines but i could be wrong.

File: vms/platformvm/state/state.go [hunk 12]
```diff
@@ -2349,11 +2442,11 @@ func (s *state) writeMetadata() error {
 }
 
 // Returns the block, status of the block, and whether it is a [stateBlk].
-// Invariant: blkBytes is safe to parse with blocks.GenesisCodec
+// Invariant: blkBytes is safe to parse with block.GenesisCodec
 //
 // TODO: Remove after v1.11.x is activated
 func parseStoredBlock(blkBytes []byte) (block.Block, choices.Status, bool, error) {
-	// Attempt to parse as blocks.Block
+	// Attempt to parse as block.Block
```
Comment (joshua-kim) [inline vms/platformvm/state/state.go:2449]: Let's make this a separate PR

**Review by joshua-kim — CHANGES_REQUESTED** [review]

File: vms/platformvm/state/diff.go [hunk 13]
```diff
@@ -319,6 +310,10 @@ func (d *diff) GetSubnetOwner(subnetID ids.ID) (fx.Owner, error) {
 	return parentState.GetSubnetOwner(subnetID)
 }
 
+func (d *diff) SetSubnetOwner(subnetID ids.ID, owner fx.Owner) {
```
Comment (joshua-kim) [inline vms/platformvm/state/diff.go:313]: Can we get a UT on this?

File: vms/platformvm/state/state.go [hunk 14]
```diff
@@ -834,9 +898,14 @@ func (s *state) GetSubnetOwner(subnetID ids.ID) (fx.Owner, error) {
 		return nil, fmt.Errorf("%q %w", subnetID, errIsNotSubnet)
 	}
 
+	s.SetSubnetOwner(subnetID, subnet.Owner)
 	return subnet.Owner, nil
 }
 
+func (s *state) SetSubnetOwner(subnetID ids.ID, owner fx.Owner) {
```
Comment (joshua-kim) [inline vms/platformvm/state/state.go:905]: Can we get a UT on this? `GetSubnetOwner` is having its behavior changed and CI is passing so it looks like we're missing some coverage on `GetSubnetOwner`

File: vms/platformvm/state/state.go (hunk 9 above)
Comment (dhrubabasu) [reply vms/platformvm/state/state.go]: I made it `fxOwnerAndSize` to match `txAndStatus`. Don't like the stutter of `fxOwnerAndSizeSize()` tho :(

Comment (joshua-kim) [reply vms/platformvm/state/state.go]: Sounds good!

File: vms/platformvm/state/state.go (hunk 8 above)
Comment (dhrubabasu) [reply vms/platformvm/state/state.go]: Modified, good call.

File: vms/platformvm/state/state.go (hunk 9 above)
Comment (dhrubabasu) [reply vms/platformvm/state/state.go]: Removed fxOwnerAndSizeSize since it is always `ids.IDLen + fxOwnerAndSize.size`

File: vms/platformvm/state/state.go (hunk 7 above)
Comment (dhrubabasu) [reply vms/platformvm/state/state.go]: I agree, we'll keep it for now but I'll probably make a follow-up cleaning up the comments

File: vms/platformvm/state/state.go (hunk 11 above)
Comment (dhrubabasu) [reply vms/platformvm/state/state.go:2343]: You are correct, it is safe to remove but I'd like to keep them since it's easier to understand that this is correct

File: vms/platformvm/state/state.go [hunk 15]
```diff
@@ -573,6 +587,18 @@ func newState(
 
 	subnetBaseDB := prefixdb.New(subnetPrefix, baseDB)
 
+	subnetOwnerDB := prefixdb.New(subnetOwnerPrefix, baseDB)
+	subnetOwnerCache, err := metercacher.New[ids.ID, fxOwnerAndSize](
+		"subnet_owner_cache",
+		metricsReg,
+		cache.NewSizedLRU[ids.ID, fxOwnerAndSize](execCfg.FxOwnerCacheSize, func(id ids.ID, f fxOwnerAndSize) int {
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:594]:
```suggestion
		cache.NewSizedLRU[ids.ID, fxOwnerAndSize](execCfg.FxOwnerCacheSize, func(_ ids.ID, f fxOwnerAndSize) int {
```

File: vms/platformvm/state/state.go [hunk 16]
```diff
@@ -819,8 +849,37 @@ func (s *state) AddSubnet(createSubnetTx *txs.Tx) {
 }
 
 func (s *state) GetSubnetOwner(subnetID ids.ID) (fx.Owner, error) {
+	if owner, exists := s.subnetOwners[subnetID]; exists {
+		return owner, nil
+	}
+
+	if ownerAndSize, cached := s.subnetOwnerCache.Get(subnetID); cached {
+		if ownerAndSize.size == 0 {
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:857]: I feel like this would be a tad more clear
```suggestion
		if ownerAndSize.owner == nil {
```

File: vms/platformvm/state/state.go [hunk 17]
```diff
@@ -819,8 +849,37 @@ func (s *state) AddSubnet(createSubnetTx *txs.Tx) {
...
+		return ownerAndSize.owner, nil
+	}
+
+	ownerBytes, err := s.subnetOwnerDB.Get(subnetID[:])
+	if err == nil {
+		var owner fx.Owner
+		if _, err := block.GenesisCodec.Unmarshal(ownerBytes, &owner); err != nil {
+			return nil, err
+		}
+		s.SetSubnetOwner(subnetID, owner)
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:869]: I don't think this is correct here. I think it would cause an unnecessary DB write.
```suggestion
```

File: vms/platformvm/state/state.go [hunk 18]
```diff
@@ -834,9 +893,14 @@ func (s *state) GetSubnetOwner(subnetID ids.ID) (fx.Owner, error) {
 		return nil, fmt.Errorf("%q %w", subnetID, errIsNotSubnet)
 	}
 
+	s.SetSubnetOwner(subnetID, subnet.Owner)
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:896]: This feels a bit weird (to be performing a write on the read)... But I think it's ok

File: vms/platformvm/state/state.go (hunk 11 above)
Comment (StephenButtolph) [reply vms/platformvm/state/state.go:2343]: I think the linter is going to yell at us because we are passing references to these values into functions. While it is true that those functions don't hold a reference to them... it feels like this is more clearly correct (until go 1.22)

File: vms/platformvm/state/state.go [hunk 19]
```diff
@@ -819,8 +849,37 @@ func (s *state) AddSubnet(createSubnetTx *txs.Tx) {
...
+		return owner, nil
+	}
+	if err != database.ErrNotFound {
+		return nil, err
+	}
+
 	subnetIntf, _, err := s.GetTx(subnetID)
 	if err != nil {
+		s.subnetOwnerCache.Put(subnetID, fxOwnerAndSize{})
 		return nil, fmt.Errorf(
```
Comment (StephenButtolph) [inline vms/platformvm/state/state.go:883]: I think we should make this return `database.ErrNotFound` to be consistent with the rest of the package + the behavior when hitting a cached value (can we add a test for that as well?)

**Review by joshua-kim — APPROVED** [review]

**Comment by StephenButtolph** [comment]
> I'd introduce `Size()` method for` fx.Owner` so to avoid complicating P-chain state further.

Generally, the `fx` doesn't know _how_ it is serialized. So it isn't clear to me how we could actually do this

**Review by StephenButtolph — APPROVED** [review]
